### PR TITLE
Nadam optimizer

### PR DESCRIPTION
**Pull request by jitindua**
It has the implementation of Nadam optimizer by using a use_nesterov flag in Adam optimizer (as done in Momentum class)

Issue link: https://github.com/tensorflow/tensorflow/issues/7715

### CODE REVIEW

**Comment by drpngx** [comment]
Could you:
1. Sign the CLA
2. git pull rebase and push again

Take a look at the recent merge:
https://github.com/tensorflow/tensorflow/pull/8405

@lukaszkaiser it looks like this implementation might be more compact and easier to maintain? Rather than splitting the class into two, it modifies the current class.

**Comment by lukaszkaiser** [comment]
I fully agree. But the other CL also had RAdam. I'm not sure how to do it best to assure these are easy to maintain. This one should also have the advantage of working with LazyAdam.

**Comment by louiehelm** [comment]
@lukaszkaiser @drpngx I agree @jitindua's code looks more compact and has testing and maintenance advantages.

One of my calculations may be more efficient in the CPU version of Nadam because of how I'm using Eigen. Perhaps @jitindua could try to adapt this efficiency if you're gonna go with his PR? See code near https://github.com/tensorflow/tensorflow/pull/8405/files#diff-fee33779233fb6ff133bbb93c7830880R301

One of the subtle / critical issues I was fixing with my Nadam interface (which this can't fix) was setting saner default decay factors.

Kingma & Ba's paper on Adam only sets beta1=0.9 and beta2=0.999 to promote their pet initialization scheme (that no one uses). And if you don't use it, those factors are absurdly large + bad. I know in theory that developers could set good decay factors on their own. But in practice, 100 out of 100 developers incorrectly compensate for Adam's poorly set default decay factors by setting depressingly low learning rates to prevent it from diverging instantly.

In some sense, Radam itself is a research "discovery" about how wrong the most widely-used learning rate / decay factors in Adam are. Its gamma factor is just a more convenient knob to help change beta1 and beta2 in the correct proportions so you can properly decrease them in tandem.

How about you pull this PR, and I'll submit two thin, python-only, init-only front-ends into core (like my current Radam one) so that there's an explicit Radam call and a Nadam optimizer that doesn't have unconscionably bad default decay factors?

**Comment by lukaszkaiser** [comment]
@louiehelm Just to give you some context: I was fighting for different defaults in TF Adam since before the first release of TF. Unluckily, people protest when things become incomparable to other frameworks and papers, so here we are with bad defaults (and we probably have to stick to them in the core).

How about this: we merge @jitindua's code first (please, take a good look at it too and comment and help to make it efficient) -- and this goes into core (it's just a slight change of core API, I hope Martin will be ok with that). Then, using his changes, we merge your NAdam and RAdam into contrib/opt, also ensuring they handle sparse updates fast, same as LazyAdam which is also in contrib/opt (as you saw, Martin is not ok to have them in core API yet). Later, when everyone starts using them instead of the core Adam, we'll migrate them into core. Does that sound ok?

**Comment by ahundt** [comment]
@lukaszkaiser sorry to hijack the conversation a bit, but where can I find out about better defaults?

**Comment by lukaszkaiser** [comment]
I guess @louiehelm can tell you more. I was using large epsilon (1e-4) to stabilize for a while, then I lowered beta1 to 0.85 and beta2 to 0.997 and brought epsilon back down to 1e-7. But I doubt this is the best. @louiehelm what do you use with pure Adam?

**Comment by ahundt** [comment]
Cool thanks, also what NADAM parameters do you recommend? Perhaps that could be included in the docstring if they won't be defaults?

I know performance will vary by application but maybe there is a good place to provide reference values to reduce the searching (via hyperopt or google) required.

**Comment by louiehelm** [comment]
I've had good performance with beta1 = 0.825, beta2 = 0.99685, lr = double.

If you have a well-tuned model using Adam's stock decay params, this beta2 lets you simply double the learning rate, while still landing back in the same favorable gradient landscape you were navigating into before (but with more stability).

I didn't do the explicit math, so there might be a more precise set of beta1 / beta2 that give this benefit. Given how many people are getting comfortable with well-tuned models that hum along with Adam, maybe this is a desirable target for future default decay factors, so that we can easily migrate researchers / developers to a new standard. At very least, figuring out decay factors that do this and then mentioning them in Adam's documentation like @ahundt suggests seems wise.

I share @lukaszkaiser's intuition that epsilon should be higher than 1e-8 but never benefited from raising it myself in practice. However, that's a parameter you wouldn't see harming your results until it bites you... so I maybe just got lucky. I haven't done the proper data interrogation to see the warning signs if 1e-8 is concerningly low or not.

**Comment by drpngx** [comment]
@jitindua how much bandwidth do you have for this? Could you sign the CLA?

**Comment by lukaszkaiser** [comment]
I only have problems with epsilon when training on tasks where the accuracy gets very high, and gradients very low, sometimes close to 0. Some tasks are like this, you want to get to 99% accuracy. But then with low epsilon Adam makes this crazy large step somewhere and gets bad. I think that a good way would be instead of having `(sqrt(v_t) + epsilon)` to have something smarter, like `max(sqrt(v_t), epsilon)` but maybe a little smoothed. So one could set epsilon higher without effecting all the normal gradients. Not sure if it's possible though, but I'd love not to have to care about that...

**Comment by martinwicke** [comment]
@lukaszkaiser @jitindua What's the verdict here?

**Comment by lukaszkaiser** [comment]
Again, here is my suggestion:

How about this: we merge @jitindua's code first (please, take a good look at it too and comment and help to make it efficient) -- and this goes into core (it's just a slight change of core API, I hope Martin will be ok with that). Then, using his changes, we merge your NAdam and RAdam into contrib/opt, also ensuring they handle sparse updates fast, same as LazyAdam which is also in contrib/opt (as you saw, Martin is not ok to have them in core API yet). Later, when everyone starts using them instead of the core Adam, we'll migrate them into core.

Do we all agree on that general plan?

**Comment by louiehelm** [comment]
@jitindua 's code is cool and all. But let's not forget that my PR still has a signed CLA, is properly based, linted, and ready to merge.

Remember those 12 hours or so it was live? Good times!

@martinwicke -- why don't we just add Nadam and Radam to golden and go back to those good dope days?

I was all set to merge down the repeated C++ for another PR when my Nadam code got unpulled  💔
https://github.com/tensorflow/tensorflow/pull/8405#issuecomment-293402921

Still game to clean up TF's C++ optimizers or even migrate some of their interfaces towards Python eventually. Can I have Nadam and Radam in core in exchange for properly refactoring all the Optimizer interfaces?

**Comment by drpngx** [comment]
Sorry @jitindua but if you don't sign the CLA there's nothing we can do. Also, if you don't have cycles to respond then I'm not sure we can really help.

@louiehelm we have made the decision that the code should live in contrib first, as we usually do. Contrib is our canary. Code that lives there has time mature with a lot of freedom.
As long as the API doesn't change, we can refactor code and make it cleaner.
As I understand, this PR has the advantage that there is less replicated code. If you think you can resubmit your PR, in contrib, with more shared code, then I think that might be a good avenue and we can drop this PR.

**Comment by vrv** [comment]
It looks like @jitindua has not responded since the original PR, and it's been 3 weeks.  So I'm going to close this PR and hope that it gets opened by someone (anyone!) who can sign the CLA, preferably under 'contrib', and with the suggestions / improvements posted here!
